core/src/main/java/com/linecorp/armeria/common/stream/StreamWriterOutputStream.java

```diff
@@ -0,0 +1,125 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+final class StreamWriterOutputStream<T> extends OutputStream {
+
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private ByteBuf buffer;
+    private final int maxBufferSize;
+    private boolean closed;
+
+    StreamWriterOutputStream(StreamWriter<T> writer,
+                             Function<? super HttpData, ? extends T> httpDataConverter,
+                             int maxBufferSize) {
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        checkArgument(maxBufferSize > 0, "maxBufferSize: %s (expected: > 0)", maxBufferSize);
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+        this.maxBufferSize = maxBufferSize;
+        buffer = Unpooled.buffer(maxBufferSize, maxBufferSize);
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        maybeDrain();
+        buffer.writeByte(b);
+    }
+
+    @Override
+    public void write(byte[] bytes, int off, int len) throws IOException {
+        requireNonNull(bytes, "bytes");
+        checkArgument(off >= 0, "off: %s (expected: >= 0)", off);
+        checkArgument(len >= 0, "len: %s (expected: >= 0)", len);
+        while (len > 0) {
+            ensureOpen();
+            maybeDrain();
+            final int writableBytes = maxBufferSize - buffer.readableBytes();
+            if (writableBytes < len) {
+                buffer.writeBytes(bytes, off, writableBytes);
+                off += writableBytes;
+                len -= writableBytes;
+            } else {
+                buffer.writeBytes(bytes, off, len);
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        ensureOpen();
+        drain();
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        flush();
+        writer.close();
+        closed = true;
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+        if (!writer.isOpen()) {
+            closed = true;
+            throw new IOException("Stream closed");
```

Comment:
`new IOException("Stream closed")` could be deduplicated.

Reply:
✅ updated